bindings/serialize-v8.cc

```diff
@@ -0,0 +1,205 @@
+      Sample s = Sample({stack.begin(), stack.end()}, vals, labels);
+      samples.push_back(s);
+    }
+    return samples;
+  }
+};
+
+class TimeNode : public Node {
+ private:
+  const CpuProfileNode *node;
+  int samplingIntervalMicros;
+
+ public:
+  TimeNode(const CpuProfileNode *node, int samplingIntervalMicros) {
+    this->node = node;
+    this->samplingIntervalMicros = samplingIntervalMicros;
+  }
+
+  virtual std::string name() {
+    return *String::Utf8Value(node->GetFunctionName());
+  }
+
+  virtual std::string filename() {
+    return *String::Utf8Value(node->GetScriptResourceName());
+  }
+
+  virtual int64_t getFileID() { return node->GetScriptId(); }
+
+  virtual int64_t lineNumber() { return node->GetLineNumber(); }
+
+  virtual int64_t columnNumber() { return node->GetColumnNumber(); }
+
+  virtual std::vector<Sample> samples(const std::deque<uint64_t> &stack,
+                                      Profile *p) {
+    std::vector<Sample> samples;
+    int64_t hitCount = node->GetHitCount();
+    std::vector<int64_t> vals = {hitCount, hitCount * samplingIntervalMicros};
+    Sample s = Sample({stack.begin(), stack.end()}, vals, std::vector<Label>());
+    samples.push_back(s);
+    return samples;
+  }
+};
+
+struct TimeEntry {
+  const CpuProfileNode *node;
+
+  // number of entries which would need to be removed from the stack after
+  // processing the node if the node were a leaf node.
+  int popCount;
+};
+
+std::unique_ptr<std::vector<char>> serializeTimeProfile(
+    CpuProfile *profile, int64_t samplingIntervalMicros,
+    int64_t startTimeNanos) {
+  int64_t durationNanos =
+      (profile->GetEndTime() - profile->GetStartTime()) * 1000;
+
+  Profile p = Profile("wall", "microseconds", samplingIntervalMicros,
+                      startTimeNanos, durationNanos);
+  p.addSampleType("sample", "count");
+  p.addSampleType("wall", "microseconds");
+
+  // Add root to entries
+  const CpuProfileNode *root = profile->GetTopDownRoot();
+  std::deque<TimeEntry> entries;
+  int32_t count = root->GetChildrenCount();
+  for (int32_t i = 0; i < count; i++) {
+    TimeEntry child = {root->GetChild(i), 1};
+    entries.push_back(child);
+  }
+
+  // Iterate over profile tree and serialize
+  std::deque<uint64_t> stack;
+  while (entries.size() > 0) {
+    TimeEntry entry = entries.front();
+    entries.pop_front();
+
+    std::unique_ptr<Node> node(
+        new TimeNode(entry.node, samplingIntervalMicros));
+    p.addSample(node, &stack);
+    int32_t count = entry.node->GetChildrenCount();
+    if (count == 0) {
+      for (int i = 0; i < entry.popCount; i++) {
+        stack.pop_front();
+      }
+      continue;
+    }
+    for (int32_t i = 0; i < count; i++) {
+      int popCount = 1;
+      if (i == 0) {
+        popCount += entry.popCount;
+      }
+      entries.push_front({entry.node->GetChild(i), popCount});
+    }
+  }
+
+  // serialize profile
+  std::vector<char> *b = new std::vector<char>();
```

Comment:
`buffer` would be a more descriptive name than `b`.

Reply:
done.